src/message/fetchActions.js

```diff
@@ -222,6 +227,42 @@ const fetchTopMostNarrow = () => async (dispatch: Dispatch, getState: GetState)
   }
 };
 
+/**
+ * Calls an async function and if unsuccessful retries the call.
+ *
+ * If the function is an API call and the response has HTTP status code 4xx
+ * the error is considered unrecoverable and the exception is rethrown, to be
+ * handled further up in the call stack.
+ */
+export async function tryFetch<T>(func: () => Promise<T>): Promise<T> {
+  const MAX_TIME_MS: number = 60000;
+  const backoffMachine = new BackoffMachine();
+
+  const timeAtStart = Date.now();
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const msElapsed = Date.now() - timeAtStart;
+    // If the first attempt had a 60-second timeout, and it took 3
+    // seconds, then the second attempt should have a 57-second
+    // timeout. Etc.
+    const timeLimitMs = MAX_TIME_MS - msElapsed;
+    try {
+      return await promiseTimeout<T>(func(), timeLimitMs);
+    } catch (e) {
+      if (isClientError(e) || e instanceof TimeoutError) {
+        throw e;
+      }
+      await backoffMachine.wait();
```

Comment:
A small simplification I notice that applies already to the existing code, but might be helpful for the changes you're making: the `await backoffMachine.wait();` could be pulled out of the `catch` block, to just be at the end of the `while` body.

I think that actually makes the most conceptual sense, too -- the backoff is happening at the bottom of the loop, just before we start a new iteration.

Reply:
Makes sense.